desc/compute/bounce_integral.py

```diff
@@ -0,0 +1,961 @@
+
+    return bp1, bp2
+    # This is no longer implemented at the moment.
+    #   If the first intersect is at a non-negative derivative, that particle
+    #   may be trapped in a well outside this snapshot of the field line. If, in
+    #   addition, the last intersect is at a non-positive derivative, then we
+    #   have information to compute a bounce integral between these points.
+    #   This single bounce integral is somewhat undefined since the field typically
+    #   does not close on itself, but in some cases it can make sense to include it.
+    #   To make this integral well-defined, an approximation is made that the field
+    #   line is periodic such that ζ = knots[-1] can be interpreted as ζ = 0 so
+    #   that the distance between these bounce points is well-defined. This is fine
+    #   as long as after a transit the field line begins physically close to where
+    #   it began on the previous transit, for then continuity of |B| implies
+    #   |B|(knots[-1] < ζ < knots[-1] + knots[0]) is close to |B|(0 < ζ < knots[0]).
+    #   We don't need to check conditions for the latter, because if they are not
+    #   satisfied, the quadrature will evaluate √(1 − λ |B|) as nan automatically.
+
+
+def _compute_bp_if_given_pitch(
+    knots, B_c, B_z_ra_c, pitch, check, *original, err=False
+):
+    """Conditionally return the ingredients needed to compute bounce integrals.
+
+    Parameters
+    ----------
+    original : tuple
+        Whatever this method returned earlier.
+    err : bool
+        Whether to raise an error if ``pitch`` is None and ``original`` is empty.
+
+    """
+    if pitch is None:
+        if err and not original:
+            raise ValueError("No pitch values were given.")
+        return original
+    else:
+        pitch = jnp.atleast_2d(pitch)
+        return *bounce_points(knots, B_c, B_z_ra_c, pitch, check), pitch
+
+
+def tanh_sinh_quad(resolution=7):
+    """Tanh-Sinh quadrature.
+
+    This function outputs the quadrature points xₖ and weights wₖ
+    for a tanh-sinh quadrature.
+
+    ∫₋₁¹ f(x) dx = ∑ₖ wₖ f(xₖ)
+
+    Parameters
+    ----------
+    resolution: int
+        Number of quadrature points, preferably odd.
+
+    Returns
+    -------
+    x : numpy array
+        Quadrature points
+    w : numpy array
+        Quadrature weights
+
+    """
+    # boundary of integral
+    x_max = jnp.array(1.0)
+    # subtract machine epsilon with buffer for floating point error
+    x_max = x_max - 10 * jnp.finfo(x_max).eps
+    # inverse of tanh-sinh transformation
+    t_max = jnp.arcsinh(2 * jnp.arctanh(x_max) / jnp.pi)
+    kh = jnp.linspace(-t_max, t_max, resolution)
+    h = 2 * t_max / (resolution - 1)
+    x = jnp.tanh(0.5 * jnp.pi * jnp.sinh(kh))
+    w = 0.5 * jnp.pi * h * jnp.cosh(kh) / jnp.cosh(0.5 * jnp.pi * jnp.sinh(kh)) ** 2
+    return x, w
+
+
+_interp1d_vec = jnp.vectorize(
+    interp1d,
+    signature="(m),(n),(n)->(m)",
+    excluded={"method", "derivative", "extrap", "period"},
```

Comment:
Referencing this: https://github.com/PlasmaControl/DESC/pull/854#issuecomment-2046031265,
just change `excluded={"method", "derivative", "extrap", "period"}` to `excluded={3}`.